File: mobile/app/src/main/java/ke/co/baziqhue/umd/Discover.kt
package ke.co.baziqhue.umd

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/** One discoverable video from YouTube. */
data class DiscoverItem(
    val videoId: String,
    val title: String,
    val channel: String,
    val thumb: String,
    val durationSec: Int,
) {
    val url get() = "https://www.youtube.com/watch?v=$videoId"
}

/** A channel hit from a Discover search — tap to browse its whole catalogue (yt-dlp). */
data class DiscoverChannel(val channelId: String, val title: String, val thumb: String) {
    val url get() = "https://www.youtube.com/channel/$channelId"
}

/** A playlist hit — tap to open in the Channel tab and bulk-download it (yt-dlp). */
data class DiscoverPlaylist(val playlistId: String, val title: String, val thumb: String) {
    val url get() = "https://www.youtube.com/playlist?list=$playlistId"
}

/** Mixed search results: videos to download + channels & playlists to open & browse. */
data class SearchResults(
    val videos: List<DiscoverItem>,
    val channels: List<DiscoverChannel>,
    val playlists: List<DiscoverPlaylist>,
)

/**
 * YouTube content discovery via the free **Data API v3**. Powers the Discover
 * shelves (Trending in Kenya / Worldwide, Music, Picks-for-you, More-from-artist).
 *
 * The API key is **embedded at build time** (secret.properties → BuildConfig) and is
 * deliberately NOT user-visible or configurable — Discover just works, like a native
 * feature. Responses are cached on disk (trending 6h, the quota-heavy search 24h) so
 * the shared key's daily quota lasts; trending changes slowly so this is plenty fresh.
 */
object Discover {

    private const val BASE = "https://www.googleapis.com/youtube/v3"
    private const val TRENDING_TTL = 6L * 60 * 60 * 1000          // 6h
    private const val SEARCH_TTL = 24L * 60 * 60 * 1000           // 24h (search costs 100 units)

    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    /**
     * ALL embedded keys, in priority order, for quota failover. A POOL
     * (YOUTUBE_API_KEYS, comma/space separated) is preferred — one key per SEPARATE
     * Google Cloud project adds real capacity (quota is per-project) — and falls back
     * to the single YOUTUBE_API_KEY. De-duped. Not stored in prefs, not shown, not editable.
     */
    @Volatile private var keysCache: List<String>? = null
    fun apiKeys(): List<String> {
        keysCache?.let { return it }
        val pool = BuildConfig.YOUTUBE_API_KEYS.split(Regex("[,\\s]+"))
            .map { it.trim() }.filter { it.isNotBlank() }
        val single = BuildConfig.YOUTUBE_API_KEY.trim()
        val all = (if (pool.isNotEmpty()) pool
                   else if (single.isNotBlank()) listOf(single) else emptyList()).distinct()
        keysCache = all
        return all
    }

    /** The first key (kept for any caller that just needs "a" key). */
    fun apiKey(): String = apiKeys().firstOrNull().orEmpty()
    fun hasKey(): Boolean = apiKeys().isNotEmpty()

    // --- cache ---------------------------------------------------------------- #
    private fun cacheDir(): File = File(Storage.aiDir(), "discover_cache").apply { if (!exists()) mkdirs() }
    private fun cacheFile(k: String) = File(cacheDir(), k.replace(Regex("[^a-zA-Z0-9]"), "_") + ".json")

    /** Cached body if present and within [ttlMs]; [ignoreTtl] returns even stale cache. */
    private fun readCache(k: String, ttlMs: Long, ignoreTtl: Boolean): String? {
        val f = cacheFile(k)
        if (!f.exists()) return null
        return try {
            val o = JSONObject(f.readText())
            if (ignoreTtl || System.currentTimeMillis() - o.optLong("ts") < ttlMs)
                o.optString("body") else null
        } catch (_: Exception) { null }
    }

    private fun writeCache(k: String, body: String) {
        try {
            cacheFile(k).writeText(JSONObject().put("ts", System.currentTimeMillis()).put("body", body).toString())
        } catch (_: Exception) {}
    }

    /** A 4xx the caller shouldn't retry (bad request / forbidden). */
    private open class PermanentHttp(message: String) : IOException(message)
    /** 403 daily-limit on THIS key — the caller rotates to the next key in the pool. */
    private class QuotaExceeded(message: String) : PermanentHttp(message)

    @Volatile private var rr = 0   // round-robin cursor — spreads load across projects

    /**
     * GET with key rotation + transient retry. Each call starts at a different key
     * (round-robin) so multiple projects share the traffic; on a 403 (daily quota)
     * it falls over to the next key, so one exhausted project never breaks Discover.
     * [url] must NOT already contain a key= param — this appends it per attempt, so
     * the on-disk cache key (which never includes the key) stays stable across keys.
     * With a single key this behaves exactly as before.
     */
    private fun httpGet(url: String): String {
        val keys = apiKeys().ifEmpty { listOf("") }
        val n = keys.size
        rr = (rr + 1) % n
        var last: Exception? = null
        for (i in 0 until n) {
            val key = keys[(rr + i) % n]
            try {
                return httpGetOnce(if (key.isBlank()) url else "$url&key=$key")
            } catch (e: QuotaExceeded) {
                last = e                                      // this key is spent → try next
            }
        }
        throw last ?: QuotaExceeded("YouTube daily limit reached on all keys.")
    }

    /**
     * One key's request, with automatic retry for *transient* failures (no network
     * yet, DNS not resolved, timeouts, 5xx, rate-limit). This kills the old "search
     * shows an error, toggle a filter and suddenly it works" flakiness. A 403 becomes
     * [QuotaExceeded] so the caller rotates keys; other permanent 4xx fail fast.
     */
    private fun httpGetOnce(url: String): String {
        var last: Exception? = null
        for (attempt in 0 until 3) {
            try {
                val req = Request.Builder().url(url).get().build()
                client.newCall(req).execute().use { resp ->
                    val body = resp.body?.string().orEmpty()
                    if (resp.isSuccessful) return body
                    val msg = parseErr(body, resp.code)
                    if (resp.code == 403) throw QuotaExceeded(msg)       // spent key → rotate
                    if (resp.code in 400..499 && resp.code != 408 && resp.code != 429)
                        throw PermanentHttp(msg)              // permanent → don't retry
                    last = IOException(msg)                   // 5xx / 408 / 429 → retry
                }
            } catch (e: PermanentHttp) {
                throw e
            } catch (e: IOException) {
                last = e                                      // connection-level failure → retry
            }
            if (attempt < 2) try { Thread.sleep(500L * (attempt + 1)) } catch (_: InterruptedException) {}
        }
        throw last ?: IOException("Couldn't reach YouTube.")
    }

    private fun parseErr(body: String, code: Int): String = try {
        if (code == 403) "YouTube daily limit reached — try again later."
        else JSONObject(body).optJSONObject("error")?.optString("message").orEmpty()
            .ifBlank { "Couldn't reach YouTube (HTTP $code)." }
    } catch (_: Exception) { "Couldn't reach YouTube (HTTP $code)." }

    /** Cache (fresh) or network; on failure fall back to stale cache. */
    private fun cachedFetch(cacheKey: String, url: String, ttlMs: Long): String {
        readCache(cacheKey, ttlMs, ignoreTtl = false)?.let { return it }
        return try {
            httpGet(url).also { writeCache(cacheKey, it) }
        } catch (e: Exception) {
            readCache(cacheKey, ttlMs, ignoreTtl = true) ?: throw e
        }
    }

    /** Most-popular videos for a region (1 quota unit). [categoryId] "10" = Music. */
    suspend fun trending(regionCode: String, categoryId: String? = null): Result<List<DiscoverItem>> =
        withContext(Dispatchers.IO) {
            if (!hasKey()) return@withContext Result.failure(IOException("Discover is unavailable."))
            try {
                val cat = if (categoryId != null) "&videoCategoryId=$categoryId" else ""
                val url = "$BASE/videos?part=snippet,contentDetails&chart=mostPopular&maxResults=20" +
                    "&regionCode=$regionCode$cat"
                Result.success(parseVideos(cachedFetch("trending_${regionCode}_${categoryId ?: "all"}", url, TRENDING_TTL)))
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    /** Search videos by name (100 quota units — cached 24h). [order] = date | relevance. */
    suspend fun search(query: String, order: String = "date"): Result<List<DiscoverItem>> =
        withContext(Dispatchers.IO) {
            if (!hasKey()) return@withContext Result.failure(IOException("Discover is unavailable."))
            if (query.isBlank()) return@withContext Result.success(emptyList())
            try {
                val q = URLEncoder.encode(query, "UTF-8")
                val url = "$BASE/search?part=snippet&type=video&order=$order&maxResults=20&q=$q"
                Result.success(parseSearch(cachedFetch("search_${order}_$query", url, SEARCH_TTL)))
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    /**
     * Mixed keyword search (videos + channels in ONE call — 100 units, cached 24h).
     * Lets Discover act as a download-first "mini YouTube": find a video to grab, or
     * a channel to open and browse its full catalogue via the existing Channel tab.
     */
    suspend fun searchMixed(query: String, order: String = "relevance"): Result<SearchResults> =
        withContext(Dispatchers.IO) {
            if (!hasKey()) return@withContext Result.failure(IOException("Discover is unavailable."))
            if (query.isBlank()) return@withContext Result.success(SearchResults(emptyList(), emptyList(), emptyList()))
            try {
                val q = URLEncoder.encode(query, "UTF-8")
                // No `type` → YouTube returns a mix of videos / channels / playlists.
                val url = "$BASE/search?part=snippet&maxResults=25&order=$order&q=$q"
                Result.success(parseMixed(cachedFetch("searchmix_${order}_$query", url, SEARCH_TTL)))
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    /** Latest uploads from a channel (2 quota units, cached 6h) — for "New from <artist>". */
    suspend fun latestUploads(channelId: String): Result<List<DiscoverItem>> =
        withContext(Dispatchers.IO) {
            if (!hasKey()) return@withContext Result.failure(IOException("Discover is unavailable."))
            try {
                val chUrl = "$BASE/channels?part=contentDetails&id=$channelId"
                val uploads = JSONObject(cachedFetch("chmeta_$channelId", chUrl, TRENDING_TTL))
                    .optJSONArray("items")?.optJSONObject(0)
                    ?.optJSONObject("contentDetails")?.optJSONObject("relatedPlaylists")
                    ?.optString("uploads").orEmpty()
                if (uploads.isBlank()) return@withContext Result.success(emptyList())
                val plUrl = "$BASE/playlistItems?part=snippet&maxResults=12&playlistId=$uploads"
                Result.success(parsePlaylistItems(cachedFetch("uploads_$channelId", plUrl, TRENDING_TTL)))
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    private fun parseMixed(body: String): SearchResults {
        val arr = JSONObject(body).optJSONArray("items")
            ?: return SearchResults(emptyList(), emptyList(), emptyList())
        val vids = ArrayList<DiscoverItem>()
        val chans = ArrayList<DiscoverChannel>()
        val plays = ArrayList<DiscoverPlaylist>()
        for (i in 0 until arr.length()) {
            val it = arr.optJSONObject(i) ?: continue
            val idObj = it.optJSONObject("id") ?: continue
            val kind = idObj.optString("kind")
            val sn = it.optJSONObject("snippet") ?: continue
            val thumb = thumbUrl(sn.optJSONObject("thumbnails"))
            when {
                kind.endsWith("video") -> idObj.optString("videoId").takeIf { it.isNotBlank() }?.let {
                    vids.add(DiscoverItem(it, sn.optString("title"), sn.optString("channelTitle"), thumb, 0))
                }
                kind.endsWith("channel") -> idObj.optString("channelId").takeIf { it.isNotBlank() }?.let {
                    chans.add(DiscoverChannel(it, sn.optString("title"), thumb))
                }
                kind.endsWith("playlist") -> idObj.optString("playlistId").takeIf { it.isNotBlank() }?.let {
                    plays.add(DiscoverPlaylist(it, sn.optString("title"), thumb))
                }
            }
        }
        return SearchResults(vids, chans, plays)
    }

    private fun parsePlaylistItems(body: String): List<DiscoverItem> {
        val arr = JSONObject(body).optJSONArray("items") ?: return emptyList()
        val out = ArrayList<DiscoverItem>(arr.length())
        for (i in 0 until arr.length()) {
            val sn = arr.optJSONObject(i)?.optJSONObject("snippet") ?: continue
            val vid = sn.optJSONObject("resourceId")?.optString("videoId").orEmpty()
            if (vid.isBlank()) continue
            out.add(DiscoverItem(vid, sn.optString("title"), sn.optString("videoOwnerChannelTitle")
                .ifBlank { sn.optString("channelTitle") }, thumbUrl(sn.optJSONObject("thumbnails")), 0))
        }
        return out
    }

    // --- parsing -------------------------------------------------------------- #
    private fun thumbUrl(thumbs: JSONObject?): String {
        if (thumbs == null) return ""
        for (k in listOf("high", "medium", "default")) {
            thumbs.optJSONObject(k)?.optString("url")?.takeIf { it.isNotBlank() }?.let { return it }
        }
        return ""
    }

    private fun parseVideos(body: String): List<DiscoverItem> {
        val arr = JSONObject(body).optJSONArray("items") ?: return emptyList()
        val out = ArrayList<DiscoverItem>(arr.length())
        for (i in 0 until arr.length()) {
            val it = arr.optJSONObject(i) ?: continue
            val id = it.optString("id")
            val sn = it.optJSONObject("snippet") ?: continue
            if (id.isBlank()) continue
            out.add(DiscoverItem(
                id, sn.optString("title"), sn.optString("channelTitle"),
                thumbUrl(sn.optJSONObject("thumbnails")),
                parseDuration(it.optJSONObject("contentDetails")?.optString("duration"))))
        }
        return out
    }

    private fun parseSearch(body: String): List<DiscoverItem> {
        val arr = JSONObject(body).optJSONArray("items") ?: return emptyList()
        val out = ArrayList<DiscoverItem>(arr.length())
        for (i in 0 until arr.length()) {
            val it = arr.optJSONObject(i) ?: continue
            val id = it.optJSONObject("id")?.optString("videoId").orEmpty()
            val sn = it.optJSONObject("snippet") ?: continue
            if (id.isBlank()) continue
            out.add(DiscoverItem(
                id, sn.optString("title"), sn.optString("channelTitle"),
                thumbUrl(sn.optJSONObject("thumbnails")), 0))
        }
        return out
    }

    /** ISO-8601 duration ("PT3M16S") -> seconds. */
    private fun parseDuration(iso: String?): Int {
        if (iso.isNullOrBlank()) return 0
        val m = Regex("PT(?:(\\d+)H)?(?:(\\d+)M)?(?:(\\d+)S)?").find(iso) ?: return 0
        val h = m.groupValues[1].toIntOrNull() ?: 0
        val mn = m.groupValues[2].toIntOrNull() ?: 0
        val s = m.groupValues[3].toIntOrNull() ?: 0
        return h * 3600 + mn * 60 + s
    }
}

File: discover.py
"""
Discover — YouTube content discovery for the desktop app (Data API v3).
=======================================================================
A faithful port of the mobile app's Discover.kt: the SAME endpoints, params,
caching (trending 6h, search 24h), transient-retry, and parsing — so the desktop
Discover behaves exactly like the phone's. Powers the Discover tab: Trending
(Kenya / Worldwide / Music), mixed search (videos + channels + playlists), and a
channel's latest uploads — all download-first via the existing engine.

The API key is EMBEDDED (not user-configurable), mirroring the mobile build. It
is read from, in order: env UMD_YOUTUBE_KEY -> youtube.key (app dir or config
dir) -> mobile/secret.properties (dev fallback). For the frozen exe, ship a
youtube.key next to it (gitignored, like secret.key).
"""

import json
import os
import re
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from pathlib import Path

import licensing

HERE = Path(__file__).resolve().parent
BASE = "https://www.googleapis.com/youtube/v3"
TRENDING_TTL = 6 * 60 * 60        # 6h (seconds)
SEARCH_TTL = 24 * 60 * 60         # 24h (search costs 100 quota units)


# --------------------------------------------------------------------------- #
#  Models
# --------------------------------------------------------------------------- #
@dataclass
class DiscoverItem:
    video_id: str
    title: str
    channel: str
    thumb: str
    duration_sec: int = 0

    @property
    def url(self):
        return f"https://www.youtube.com/watch?v={self.video_id}"

    def as_dict(self):
        return {"video_id": self.video_id, "title": self.title, "channel": self.channel,
                "thumb": self.thumb, "duration_sec": self.duration_sec, "url": self.url}


@dataclass
class DiscoverChannel:
    channel_id: str
    title: str
    thumb: str

    @property
    def url(self):
        return f"https://www.youtube.com/channel/{self.channel_id}"

    def as_dict(self):
        return {"channel_id": self.channel_id, "title": self.title, "thumb": self.thumb, "url": self.url}


@dataclass
class DiscoverPlaylist:
    playlist_id: str
    title: str
    thumb: str

    @property
    def url(self):
        return f"https://www.youtube.com/playlist?list={self.playlist_id}"

    def as_dict(self):
        return {"playlist_id": self.playlist_id, "title": self.title, "thumb": self.thumb, "url": self.url}


# --------------------------------------------------------------------------- #
#  API key (embedded, like the mobile build)
# --------------------------------------------------------------------------- #
_keys_cache = None


def _from_secret_properties():
    f = HERE / "mobile" / "secret.properties"
    if not f.is_file():
        return ""
    try:
        for line in f.read_text(encoding="utf-8").splitlines():
            s = line.strip()
            if s.startswith("YOUTUBE_API_KEY") and "=" in s:
                return s.split("=", 1)[1].strip()
    except OSError:
        pass
    return ""


def api_keys():
    """ALL configured YouTube keys, in priority order, for quota failover.
    Sources: env UMD_YOUTUBE_KEY (comma/space separated) -> youtube.key (ONE KEY
    PER LINE; '#' lines ignored) -> mobile/secret.properties. De-duped, ordered.
    Add more keys = paste more lines into youtube.key; rotation is automatic."""
    global _keys_cache
    if _keys_cache is not None:
        return _keys_cache
    out = []
    env = (os.environ.get("UMD_YOUTUBE_KEY") or "").strip()
    if env:
        out += [k.strip() for k in re.split(r"[,\s]+", env) if k.strip()]
    if not out:
        for p in (HERE / "youtube.key", licensing.config_dir() / "youtube.key"):
            try:
                if p.is_file():
                    out += [ln.strip() for ln in p.read_text(encoding="utf-8").splitlines()
                            if ln.strip() and not ln.strip().startswith("#")]
                    if out:
                        break
            except OSError:
                pass
    if not out:
        sp = _from_secret_properties()
        if sp:
            out.append(sp)
    seen, uniq = set(), []
    for k in out:
        if k not in seen:
            seen.add(k)
            uniq.append(k)
    _keys_cache = uniq
    return uniq


def api_key():
    ks = api_keys()
    return ks[0] if ks else ""


def has_key():
    return bool(api_keys())


# --------------------------------------------------------------------------- #
#  Disk cache (mirrors the Kotlin cache: {ts, body} JSON per key)
# --------------------------------------------------------------------------- #
def _cache_dir():
    d = licensing.config_dir() / "discover_cache"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _cache_file(key):
    return _cache_dir() / (re.sub(r"[^a-zA-Z0-9]", "_", key) + ".json")


def _read_cache(key, ttl, ignore_ttl):
    f = _cache_file(key)
    if not f.is_file():
        return None
    try:
        o = json.loads(f.read_text(encoding="utf-8"))
        if ignore_ttl or (time.time() - o.get("ts", 0)) < ttl:
            return o.get("body", "")
    except Exception:
        return None
    return None


def _write_cache(key, body):
    try:
        _cache_file(key).write_text(json.dumps({"ts": time.time(), "body": body}), encoding="utf-8")
    except OSError:
        pass


# --------------------------------------------------------------------------- #
#  HTTP with retry for transient failures (port of the Kotlin retry)
# --------------------------------------------------------------------------- #
class _PermanentHttp(Exception):
    """A 4xx the caller shouldn't retry (bad request / forbidden)."""


class _QuotaExceeded(_PermanentHttp):
    """403 quota/limit — rotate to the next API key if we have one."""


def _parse_err(body, code):
    try:
        if code == 403:
            return "YouTube daily limit reached — try again later."
        msg = ((json.loads(body) or {}).get("error", {}) or {}).get("message", "")
        return msg or f"Couldn't reach YouTube (HTTP {code})."
    except Exception:
        return f"Couldn't reach YouTube (HTTP {code})."


def _swap_key(url, key):
    """Put `key` into the URL's key= param (keys rotate; the rest of the URL —
    and therefore the cache key — stays the same)."""
    if not key:
        return url
    if re.search(r"[?&]key=", url):
        return re.sub(r"([?&]key=)[^&]*", lambda m: m.group(1) + key, url)
    return url + ("&" if "?" in url else "?") + "key=" + key


def _http_get_once(url):
    last = None
    for attempt in range(3):
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "UMD-Discover"})
            with urllib.request.urlopen(req, timeout=30) as resp:
                return resp.read().decode("utf-8", "replace")
        except urllib.error.HTTPError as e:
            body = ""
            try:
                body = e.read().decode("utf-8", "replace")
            except Exception:
                pass
            msg = _parse_err(body, e.code)
            if e.code == 403:
                raise _QuotaExceeded(msg)             # spent key → caller rotates
            if 400 <= e.code < 500 and e.code not in (408, 429):
                raise _PermanentHttp(msg)            # permanent → don't retry
            last = IOError(msg)                       # 5xx / 408 / 429 → retry
        except (urllib.error.URLError, TimeoutError, OSError) as e:
            last = e                                  # connection-level failure → retry
        if attempt < 2:
            time.sleep(0.5 * (attempt + 1))
    raise last or IOError("Couldn't reach YouTube.")


def _http_get(url):
    """Fetch with transient-retry, AND automatic key rotation: if a key's daily
    quota is exhausted (403), fall through to the next configured key so Discover
    keeps working. With a single key this behaves exactly as before."""
    keys = api_keys() or [""]
    last = None
    for key in keys:
        try:
            return _http_get_once(_swap_key(url, key))
        except _QuotaExceeded as e:
            last = e                                  # this key is spent — try next
            continue
    raise last or _QuotaExceeded("YouTube daily limit reached on all keys.")


def _cached_fetch(cache_key, url, ttl):
    cached = _read_cache(cache_key, ttl, ignore_ttl=False)
    if cached is not None:
        return cached
    try:
        body = _http_get(url)
        _write_cache(cache_key, body)
        return body
    except Exception:
        stale = _read_cache(cache_key, ttl, ignore_ttl=True)
        if stale is not None:
            return stale
        raise


# --------------------------------------------------------------------------- #
#  Parsing
# --------------------------------------------------------------------------- #
def _thumb_url(thumbs):
    if not thumbs:
        return ""
    for k in ("high", "medium", "default"):
        u = (thumbs.get(k) or {}).get("url", "")
        if u:
            return u
    return ""


def _parse_duration(iso):
    if not iso:
        return 0
    m = re.search(r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?", iso)
    if not m:
        return 0
    h = int(m.group(1) or 0)
    mn = int(m.group(2) or 0)
    s = int(m.group(3) or 0)
    return h * 3600 + mn * 60 + s


def _parse_videos(body):
    out = []
    for it in (json.loads(body).get("items") or []):
        vid = it.get("id", "")
        sn = it.get("snippet") or {}
        if not vid or not sn:
            continue
        out.append(DiscoverItem(vid, sn.get("title", ""), sn.get("channelTitle", ""),
                                _thumb_url(sn.get("thumbnails")),
                                _parse_duration((it.get("contentDetails") or {}).get("duration"))))
    return out


def _parse_search(body):
    out = []
    for it in (json.loads(body).get("items") or []):
        vid = (it.get("id") or {}).get("videoId", "")
        sn = it.get("snippet") or {}
        if not vid or not sn:
            continue
        out.append(DiscoverItem(vid, sn.get("title", ""), sn.get("channelTitle", ""),
                                _thumb_url(sn.get("thumbnails")), 0))
    return out


def _parse_mixed(body):
    vids, chans, plays = [], [], []
    for it in (json.loads(body).get("items") or []):
        ido = it.get("id") or {}
        kind = ido.get("kind", "")
        sn = it.get("snippet") or {}
        if not sn:
            continue
        thumb = _thumb_url(sn.get("thumbnails"))
        if kind.endswith("video") and ido.get("videoId"):
            vids.append(DiscoverItem(ido["videoId"], sn.get("title", ""), sn.get("channelTitle", ""), thumb, 0))
        elif kind.endswith("channel") and ido.get("channelId"):
            chans.append(DiscoverChannel(ido["channelId"], sn.get("title", ""), thumb))
        elif kind.endswith("playlist") and ido.get("playlistId"):
            plays.append(DiscoverPlaylist(ido["playlistId"], sn.get("title", ""), thumb))
    return {"videos": vids, "channels": chans, "playlists": plays}


def _parse_playlist_items(body):
    out = []
    for it in (json.loads(body).get("items") or []):
        sn = it.get("snippet") or {}
        vid = (sn.get("resourceId") or {}).get("videoId", "")
        if not vid:
            continue
        ch = sn.get("videoOwnerChannelTitle") or sn.get("channelTitle", "")
        out.append(DiscoverItem(vid, sn.get("title", ""), ch, _thumb_url(sn.get("thumbnails")), 0))
    return out


# --------------------------------------------------------------------------- #
#  Public API (mirrors Discover.kt)
# --------------------------------------------------------------------------- #
def trending(region_code, category_id=None):
    """Most-popular videos for a region (1 quota unit). category_id '10' = Music."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    cat = f"&videoCategoryId={category_id}" if category_id else ""
    url = (f"{BASE}/videos?part=snippet,contentDetails&chart=mostPopular&maxResults=20"
           f"&regionCode={region_code}{cat}&key={api_key()}")
    return _parse_videos(_cached_fetch(f"trending_{region_code}_{category_id or 'all'}", url, TRENDING_TTL))


def search(query, order="date"):
    """Search videos by name (100 quota units, cached 24h). order = date | relevance."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    if not query.strip():
        return []
    q = urllib.parse.quote(query)
    url = f"{BASE}/search?part=snippet&type=video&order={order}&maxResults=20&q={q}&key={api_key()}"
    return _parse_search(_cached_fetch(f"search_{order}_{query}", url, SEARCH_TTL))


def search_mixed(query, order="relevance"):
    """Mixed keyword search — videos + channels + playlists in ONE call (cached 24h)."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    if not query.strip():
        return {"videos": [], "channels": [], "playlists": []}
    q = urllib.parse.quote(query)
    url = f"{BASE}/search?part=snippet&maxResults=25&order={order}&q={q}&key={api_key()}"
    return _parse_mixed(_cached_fetch(f"searchmix_{order}_{query}", url, SEARCH_TTL))


def latest_uploads(channel_id):
    """A channel's latest uploads (2 quota units, cached 6h) — for artist discovery."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    ch_url = f"{BASE}/channels?part=contentDetails&id={channel_id}&key={api_key()}"
    meta = json.loads(_cached_fetch(f"chmeta_{channel_id}", ch_url, TRENDING_TTL))
    items = meta.get("items") or []
    uploads = ""
    if items:
        uploads = (((items[0].get("contentDetails") or {}).get("relatedPlaylists") or {}).get("uploads", ""))
    if not uploads:
        return []
    pl_url = f"{BASE}/playlistItems?part=snippet&maxResults=12&playlistId={uploads}&key={api_key()}"
    return _parse_playlist_items(_cached_fetch(f"uploads_{channel_id}", pl_url, TRENDING_TTL))


def playlist_items(playlist_id):
    """Videos inside a playlist (1 quota unit, cached 6h) — open & bulk-download."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    url = f"{BASE}/playlistItems?part=snippet&maxResults=25&playlistId={playlist_id}&key={api_key()}"
    return _parse_playlist_items(_cached_fetch(f"plitems_{playlist_id}", url, TRENDING_TTL))


def channel_info(channel_id):
    """Channel header: title, image, subscriber & video counts, uploads playlist
    (1 quota unit, cached 6h). Powers the Discover channel view."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    url = f"{BASE}/channels?part=snippet,statistics,contentDetails&id={channel_id}&key={api_key()}"
    items = json.loads(_cached_fetch(f"chinfo_{channel_id}", url, TRENDING_TTL)).get("items") or []
    if not items:
        return None
    it = items[0]
    sn = it.get("snippet") or {}
    stt = it.get("statistics") or {}
    return {
        "id": channel_id,
        "title": sn.get("title", ""),
        "thumb": _thumb_url(sn.get("thumbnails")),
        "subs": None if stt.get("hiddenSubscriberCount") else stt.get("subscriberCount"),
        "videos": stt.get("videoCount"),
        "uploads": (((it.get("contentDetails") or {}).get("relatedPlaylists") or {}).get("uploads", "")),
    }


def channel_playlists(channel_id):
    """A channel's public playlists (1 quota unit, cached 6h)."""
    if not has_key():
        raise IOError("Discover is unavailable (no YouTube key).")
    url = (f"{BASE}/playlists?part=snippet&channelId={channel_id}&maxResults=15&key={api_key()}")
    arr = json.loads(_cached_fetch(f"chpls_{channel_id}", url, TRENDING_TTL)).get("items") or []
    out = []
    for it in arr:
        sn = it.get("snippet") or {}
        pid = it.get("id", "")
        if pid:
            out.append(DiscoverPlaylist(pid, sn.get("title", ""), _thumb_url(sn.get("thumbnails"))))
    return out
